### PR TITLE
Replace getByBlockNumber by getByBlockHeader

### LINKED ISSUE #4789: Change getByBlockNumber to be getByBlockHeader
Following https://github.com/hyperledger/besu/pull/4743

We should update all the usages we can to be getByBlockHeader and phase out getByBlockNumber.

getByBlockNumber can cause issues in a post-shanghai (TimestampSchedule-enabled) chain.

This has been intentionally separated out of #4743 so it doesn't make the PR huge, see this spike for an idea of what the changes might look like: https://github.com/hyperledger/besu/pull/4687

Another related item for clean-up:
https://github.com/hyperledger/besu/pull/4743#discussion_r1045390099

Tasks
- [x] Simple replacements of getByBlockNumber to getByBlockHeader
- [x] Create getForNextBlock (used in AbstractBlockCreator, PoWMinerExecutor and likely in BFT)
- [x] Investigate FlexiblePrivacyPrecompiledContract usage in RunnerBuilder.createPrivateTransactionObserver()
- [x] Handle Dao fork and classic fork in ProtocolScheduleBuilder
- [x] BFT sequence number usage
- [x] Change streamMilestoneBlocks to a findSchedule(Predicate) [need to fix usage in TransactionPoolFactory, RunnerBuilder]
- [x] Consider deleting getByBlockNumber usage in CombinedProtocolScheduleFactory
- [x] Fix tests to getByBlockHeader or create a test ProtocolSchedule that has getByBlockNumber inc streamMilestoneBlocks for checking milestones
- [x] ScheduleBasedBlockHeaderFunctions getBlockHeaderFunctions usage of getByBlockNumber
- [x] CheckpointSyncDownloadPipelineFactory getBlockHeaderFunctions usage of getByBlockNumber
- [x] Replace HeaderBasedProtocolSchedule with ProtocolSchedule

### PR DESCRIPTION
**Pull request by gfukushima**
## PR description
First batch that replaces getByBlockNumber by getByBlockHeader. These batch includes trivial replacements with none or minimal changes to the logic of the code base.

## Fixed Issue(s)

Partially fixes #4789

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java [hunk 1]
```diff
@@ -169,6 +167,7 @@ public void getByBlockNumber_returnsTimestampScheduleIfPresent() {
     verifyNoMergeScheduleInteractions();
   }
 
+  @Ignore
```
Comment (macfarla) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java:170]: why ignore these tests? is there a follow up task?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/DebugReplayBlock.java [hunk 2]
```diff
@@ -69,20 +69,18 @@ protected Object resultByBlockNumber(
       return new JsonRpcErrorResponse(request.getRequest().getId(), UNKNOWN_BLOCK);
     }
 
+    final Block block = maybeBlock.get();
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/DebugReplayBlock.java:72]: do we need a check for null

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastImportBlocksStepTest.java [hunk 3]
```diff
@@ -60,7 +60,7 @@ public class FastImportBlocksStepTest {
 
   @BeforeEach
   public void setUp() {
-    when(protocolSchedule.getByBlockNumber(anyLong())).thenReturn(protocolSpec);
+    when(protocolSchedule.getByBlockHeader(any(BlockHeader.class))).thenReturn(protocolSpec);
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastImportBlocksStepTest.java:63]: there's a couple of other spots where anyLong() was not replaced with this class - are we losing anything by just having any()

**Review by macfarla — APPROVED** [review]
LGTM. couple of comments

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java (hunk 1 above)
Comment (gfukushima) [reply consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java:170]: this file was prob committed accidentally. Thanks @macfarla

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/DebugReplayBlock.java (hunk 2 above)
Comment (gfukushima) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/DebugReplayBlock.java:72]: line 68 checks for that.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastImportBlocksStepTest.java (hunk 3 above)
Comment (gfukushima) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastImportBlocksStepTest.java]: Changed to use any()

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java [hunk 4]
```diff
@@ -141,6 +141,15 @@ public long headBlockNumber() {
     return blockchain.getChainHeadBlockNumber();
   }
 
+  /**
+   * Return the header of the head of the chain.
+   *
+   * @return The header of the head of the chain.
+   */
+  public BlockHeader headBlockHeader() {
```
Comment (jframe) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java:149]: Can you add a test for this

Comment (gfukushima) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java:149]: Done

**Review by jframe — APPROVED** [review]

**Review by siladu — APPROVED** [review]
LGTM

**Comment by siladu** [comment]
Bug found during goerli sync: https://github.com/hyperledger/besu/issues/5064

**Pull request merged**